core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,171 @@
+
+  public DefaultShuffleMapOutputWriter(
+      int shuffleId,
+      int mapId,
+      int numPartitions,
+      ShuffleWriteMetricsReporter metrics,
+      IndexShuffleBlockResolver blockResolver,
+      SparkConf sparkConf) {
+    this.shuffleId = shuffleId;
+    this.mapId = mapId;
+    this.numPartitions = numPartitions;
+    this.metrics = metrics;
+    this.blockResolver = blockResolver;
+    this.bufferSize =
+      (int) (long) sparkConf.get(
+        package$.MODULE$.SHUFFLE_UNSAFE_FILE_OUTPUT_BUFFER_SIZE()) * 1024;
+    this.partitionLengths = new long[numPartitions];
+    this.partitionWriters = new ShufflePartitionWriter[numPartitions];
+    this.outputFile = blockResolver.getDataFile(shuffleId, mapId);
+    this.outputTempFile = Utils.tempFileWith(outputFile);
+  }
+
+  @Override
+  public ShufflePartitionWriter getNextPartitionWriter() throws IOException {
+    initStream();
+    ShufflePartitionWriter shufflePartitionWriter =
+      new DefaultShufflePartitionWriter(
+        new DefaultShuffleBlockOutputStreamImpl());
+    partitionWriters[currPartitionId++] = shufflePartitionWriter;
+    return shufflePartitionWriter;
+  }
+
+  @Override
+  public void commitAllPartitions() throws IOException {
+    for (int pId = 0; pId < numPartitions; pId ++) {
+      partitionLengths[pId] = partitionWriters[pId].getLength();
+    }
+    blockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, outputTempFile);
+    // TODO: Maybe monitor a little more intelligably then checking files?
+    if (!outputFile.getParentFile().isDirectory() && !outputFile.getParentFile().mkdirs()) {
+      throw new IOException(
+              String.format(
+                      "Failed to create shuffle file directory at %s.",
+                      outputFile.getParentFile().getAbsolutePath()));
+    }
+    if (!outputFile.isFile() && !outputFile.createNewFile()) {
+      throw new IOException(
+              String.format(
+                      "Failed to create empty shuffle file at %s.", outputFile.getAbsolutePath()));
+    }
+  }
+
+  @Override
+  public void abort(Throwable error) throws IOException {
+    cleanUp();
+    if (!outputTempFile.delete()) {
+      log.warn("Failed to delete temporary shuffle file at {}", outputTempFile.getAbsolutePath());
+    }
+    if (!outputFile.delete()) {
+      log.warn("Failed to delete outputshuffle file at {}", outputTempFile.getAbsolutePath());
+    }
+  }
+
+  private void cleanUp() throws IOException {
+    if (outputBufferedFileStream != null) {
+      outputBufferedFileStream.flush();
+      outputBufferedFileStream.close();
+    }
+
+    if (outputFileStream != null) {
+      outputFileStream.flush();
+      outputFileStream.close();
+    }
+  }
+
+  private void initStream() throws IOException {
+    if (outputFileStream == null) {
+      outputFileStream = new FileOutputStream(outputTempFile, true);
+      ts = new TimeTrackingOutputStream(metrics, outputFileStream);
```

Comment:
why do you need a reference to `outputFileStream` and `ts`? It doesn't seem to be used for any other purpose than for constructing the `BufferedOutputStream`? I could be missing something though.

Reply:
Shouldn't we always be writing to `ts` so that we get the time tracking? But this gets potentially inconsistent when considering trying to write to the output channels.

Reply:
By writing to `outputBufferedFileStream` you get the benefits of `ts`. However, as you can see from the channel implementation. You are unable to use the this for `transferTo.` But its better to use `ts` where we can, imo.